odl/contrib/solvers/spdhg/examples/spdhg_PET_1k.py

```diff
@@ -0,0 +1,626 @@
+# Copyright 2014-2017 The ODL contributors
+#
+# This file is part of ODL.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""An example of using the SPDHG algorithm to solve a PET reconstruction
+problem with total variation prior. As we do not exploit any smoothness here
+we only expect a 1/k convergence of the ergodic sequence in a Bregman sense.
+
+Note that this example uses the astra toolbox.
+
+Reference
+---------
+Chambolle, A., Ehrhardt, M. J., Richtárik, P., & Schönlieb, C.-B. (2017).
+Stochastic Primal-Dual Hybrid Gradient Algorithm with Arbitrary Sampling and
+Imaging Applications. Retrieved from http://arxiv.org/abs/1706.04957
+"""
+
+from __future__ import print_function, division
+import odl.contrib.solvers.spdhg as spdhg
+import odl.contrib.datasets.images as images
+import matplotlib.pyplot as plt
+import matplotlib
+import numpy as np
+from scipy.ndimage.filters import gaussian_filter as sc_gaussian
+from skimage.io import imsave as sk_imsave
+import odl
+import brewer2mpl
+import os
+
+
+folder_out = '.'  # create folder for data
+nepochs = 300  # number of epochs
+#nepochs = 10  # number of epochs
+niter_target = 1000  # number of iterations for the optimal reconstruction
+nvoxelx = 100  # set problem size
+
+######## not edit needed after this line ###########
+
+filename = 'spdhg_PET_1k_{}x{}'.format(nvoxelx, nvoxelx)
+subfolder = '{}epochs'.format(nepochs)
+
+folder_main = '{}/{}'.format(folder_out, filename)
+os.makedirs(folder_main, exist_ok=True)
+
+folder_today = '{}/{}'.format(folder_main, subfolder)
+os.makedirs(folder_today, exist_ok=True)
+
+folder_npy = '{}/npy'.format(folder_today)
+os.makedirs(folder_npy, exist_ok=True)
+
+# create geometry of operator
+X = odl.uniform_discr(min_pt=[-20, -20], max_pt=[20, 20],
+                      shape=[nvoxelx, nvoxelx], dtype='float32')
+
+geometry = odl.tomo.parallel_beam_geometry(X, num_angles=200, det_shape=200)
+
+G = odl.BroadcastOperator(*[odl.tomo.RayTransform(X, gi, impl='astra_cpu')
+                            for gi in geometry])
+
+# create ground truth
+Y = G.range
+groundtruth = X.element(images.resolution_phantom(shape=X.shape))
+clim = [0, 1]
+tol_norm = 1.05
+
+# save images and data
+file_data = '{}/data.npy'.format(folder_main)
```

Comment:
This should now be loadable separately, no?

Reply:
would do you mean? the idea is that one does not need to recompute things if one executes this twice so some results and data are stored